### DISCUSSION TITLE
Making it easier to filter out completed shows

### DISCUSSION THREAD

**Discussion by limbero** [trakt/trakt-api#753] [Q&A]
I have an app that functions much like https://trakt.tv/users/me/progress, but it's always filtered to only display shows with episodes left to watch.

There are two basic ways to get the list of such shows:

1. The `extended=full` approach. The upside is it's just one network call. The downside is that call can take upwards of 15 seconds.
    1. `https://api.trakt.tv/users/me/watched/shows?extended=full`
    2. Remove shows where `show.plays == show.show.aired_episodes` (slightly different for reset shows, but all the info is returned from this call)
2. The `extended=noseasons` approach. The upside is call i is much much faster. The downside is call ii happens 160 times for my user, even though I only have 17 shows with episodes left to watch.
    1. `https://api.trakt.tv/users/me/watched/shows?extended=noseasons`
    2. For each show, call `https://api.trakt.tv/shows/:slug/progress/watched`
    3. Remove shows where `show.completed == show.aired` (slightly different for reset shows, but all the info is returned from the calls in ii)

I can't make much use of `last_updated_at`, because in the first alternative I won't see it until I've already requested all the data, and in the second alternative there might still have aired more episodes, and `/shows/:slug/progress/watched` doesn't even tell me if the show is active or has ended.

I could theoretically do 1 first for every user, and on subsequent loads do 2, but first of all that's quite messy, and second I would still have to call `/shows/:slug/progress/watched` for every show that hasn't ended, or has updated.

Would it be possible to add an `aired` field to the show object for `extended=noseasons`? Not sure what kind of database join that entails, so maybe it's not possible, or it would simply make all the calls just as slow as `extended=full`. If that's so, a `status` field would at least let me cache the `aired` number for all the shows that have ended, without combining approaches 1 and 2 or making even more network calls.

**Comment by limbero** [comment]
Another alternative / complement to that could be a `last_aired_at` field, that would be very helpful with caching.

**Comment by limbero** [comment]
Hey again, any chance this could happen? Would cut the number of calls I need to do on pageload by 90%. I thought for a second I might cache locally if a show has ended, but trouble is some shows un-end, like Will & Grace or Gilmore Girls.

**Comment by rectifyer** [comment]
What about getting extended info like `https://api.trakt.tv/users/me/watched/shows?extended=noseasons,full` -- that has `show.aired_episodes` in the response.

**Comment by limbero** [comment]
oh no kidding, I can combine multiple levels?! that's amazing, and I'm pretty sure it will do the trick!

**Comment by limbero** [comment]
it works, but is it actually faster than just `?extended=full`? It's hard to test since I pretty obviously hit caches, but it seems like they both take on average 7-10 seconds for my user.

**Comment by rectifyer** [comment]
Uncached, they would both take a little while to calculate. It's mostly doing the same things behind the scenes, but the JSON output is reduced without seasons.

**Comment by limbero** [comment]
I see. My main issue is request failures / slow loading when using my page on a phone or other weak connection. I was looking at this in an attempt to cache more, but I think waiting 7-10 seconds to know if I can skip a bunch of fast requests is probably not worth the tradeoff. Thanks for trying to help!

**Comment by JumBay** [comment]
I still think it would be interesting to filter out completed shows. My app also provide an "Up next to watch" feature and for some users `https://api.trakt.tv/users/me/watched/shows?extended=noseasons,full` will return a lot of unnecessary data. 

Simkl [handles this very well](https://simkl.docs.apiary.io/#reference/sync/get-all-items/get-all-items-in-the-user's-watchlist). Basically you ask for shows in the status you want (completed, watching, watchlist...) with the `date_from` parameter. If this parameter is omitted it will return the complete list of the wanted status, otherwise it will only return new items from this date. It's super fast and lightweight. In addition the list will contains the next episode to watch, if any, for each item.

**Comment by rysson** [comment]
The same. App folder contains next to watch, and 100% must be omitted. 

It will be very useful:
- `aired_episodes` in each season
- `aired_seasons` in the show

It allows to calculate next episode to watch. In every mode we have (fist unwatched, newest, next after last watched).
